packages/tre/src/plugins/r2/gateway.ts

```diff
@@ -42,160 +73,334 @@ export interface R2Objects {
+
+  constructor(private readonly log: Log, legacyStorage: Storage) {
+    this.#storage = legacyStorage.getNewStorage();
+    this.#storage.db.pragma("case_sensitive_like = TRUE");
+    this.#storage.db.exec(SQL_SCHEMA);
+    this.#stmts = sqlStmts(this.#storage.db);
+  }
+
+  #backgroundDelete(blobId: string) {
+    void this.#storage.blob.delete(blobId).catch(() => {});
+  }
 
   async head(key: string): Promise<R2Object> {
     validate.key(key);
 
-    // Get value, returning null if not found
-    const stored = await this.storage.head<R2ObjectMetadata>(key);
+    const row = this.#stmts.getByKey.get({ key });
+    if (row === undefined) throw new NoSuchKey();
 
-    if (stored?.metadata === undefined) throw new NoSuchKey();
-    const { metadata } = stored;
-    metadata.range = { offset: 0, length: metadata.size };
-
-    return new R2Object(metadata);
+    const range: R2Range = { offset: 0, length: row.size };
+    return new R2Object(row, range);
   }
 
   async get(
     key: string,
     options: R2GetOptions = {}
   ): Promise<R2ObjectBody | R2Object> {
-    const meta = await this.head(key);
-    validate.key(key).condition(meta, options.onlyIf);
-
-    let range = options.range ?? {};
-    if (options.rangeHeader !== undefined) {
-      const ranges = _parseRanges(options.rangeHeader, meta.size);
-      if (ranges?.length === 1) {
-        // If the header contained a single range, convert it to an R2Range.
-        // Note `start` and `end` are inclusive.
-        const [start, end] = ranges[0];
-        range = { offset: start, length: end - start + 1 };
-      } else {
-        // If the header was invalid, or contained multiple ranges, just return
-        // the full response
-        range = {};
-      }
-    }
+    validate.key(key);
 
-    let stored: RangeStoredValueMeta<R2ObjectMetadata> | undefined;
+    // TODO(soon): we may need to make this a transaction for multipart get
+    const row = this.#stmts.getByKey.get({ key });
+    if (row === undefined) throw new NoSuchKey();
 
-    try {
-      stored = await this.storage.getRange<R2ObjectMetadata>(key, range);
-    } catch {
-      throw new InvalidRange();
-    }
-    if (stored?.metadata === undefined) throw new NoSuchKey();
-    const { value, metadata } = stored;
-    // add range should it exist
-    if ("range" in stored && stored.range !== undefined) {
-      metadata.range = stored.range;
+    const defaultRange: R2Range = { offset: 0, length: row.size };
+    const meta = new R2Object(row, defaultRange);
+    const range = validate
+      .condition(meta, options.onlyIf)
+      .range(options, row.size);
+
+    assert(row.blob_id !== null); // TODO(soon): add multipart support
+    const value = await this.#storage.blob.get(row.blob_id, range);
+    if (value === null) throw new NoSuchKey();
+
+    let valueRange: R2Range | undefined;
+    if (value.range !== undefined) {
+      assert(!Array.isArray(value.range));
```

Comment:
Why do we need to assert this? Is it possible? Is that what old ranges looked like?

Reply:
Or is it just for type narrowing?